### ISSUE TITLE
get_total_spent query slow

### ISSUE THREAD

**Issue by des79** [woocommerce/woocommerce#20513]
**Describe the bug**
In my slow-query -log I can see that query executed on method get_total_spent in /includes/data-stores/class-wc-customer-data-store.php is a very slow query. Here some details:
`# Query_time: 2.862296  Lock_time: 0.000155 Rows_sent: 1  Rows_examined: 543172
SET timestamp=1528876847;
SELECT SUM(meta2.meta_value)
				FROM wp_posts as posts
				LEFT JOIN wp_postmeta AS meta ON posts.ID = meta.post_id
				LEFT JOIN wp_postmeta AS meta2 ON posts.ID = meta2.post_id
				WHERE   meta.meta_key       = '_customer_user'
				AND     meta.meta_value     = '39046'
				AND     posts.post_type     = 'shop_order'
				AND     posts.post_status   IN ( 'wc-processing','wc-completed' )
				AND     meta2.meta_key      = '_order_total';
`
On my site with +90000 customers and +8000 orders, this slow query overload my AWS Aurora DB and web server go down.

**WordPress Environment**
WP ver is 4.9.6
<details>

```
`
### WordPress Environment ###

Home URL: https://www.piuchepuoi.it
Site URL: https://www.piuchepuoi.it
WC Version: 3.3.5
Log Directory Writable: ✔
WP Version: 4.9.6
WP Multisite: –
WP Memory Limit: 512 MB
WP Debug Mode: –
WP Cron: –
Language: it_IT

### Server Environment ###

Server Info: Apache/2.2.15
PHP Version: 5.6.25
PHP Post Max Size: 8 MB
PHP Time Limit: 0
PHP Max Input Vars: 1000
cURL Version: 7.19.7
NSS/3.27.1

SUHOSIN Installed: –
MySQL Version: 5.6.10
Max Upload Size: 2 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.3.5
WC Database Prefix: wp_
MaxMind GeoIP Database: ✔
Dimensione totale database: 4311.11MB
Dimensione dati database: 2679.64MB
Dimensione indice database: 1631.47MB
wp_woocommerce_sessions: Dati: 4.03MB + Indice: 0.08MB
wp_woocommerce_api_keys: Dati: 0.02MB + Indice: 0.03MB
wp_woocommerce_attribute_taxonomies: Dati: 0.02MB + Indice: 0.02MB
wp_woocommerce_downloadable_product_permissions: Dati: 1.52MB + Indice: 2.17MB
wp_woocommerce_order_items: Dati: 1.52MB + Indice: 0.52MB
wp_woocommerce_order_itemmeta: Dati: 9.52MB + Indice: 10.81MB
wp_woocommerce_tax_rates: Dati: 0.02MB + Indice: 0.06MB
wp_woocommerce_tax_rate_locations: Dati: 0.02MB + Indice: 0.05MB
wp_woocommerce_shipping_zones: Dati: 0.02MB + Indice: 0.00MB
wp_woocommerce_shipping_zone_locations: Dati: 0.02MB + Indice: 0.03MB
wp_woocommerce_shipping_zone_methods: Dati: 0.02MB + Indice: 0.00MB
wp_woocommerce_payment_tokens: Dati: 0.02MB + Indice: 0.02MB
wp_woocommerce_payment_tokenmeta: Dati: 0.02MB + Indice: 0.03MB
wp_woocommerce_log: Dati: 0.02MB + Indice: 0.02MB
wp_advads_clicks: Dati: 0.06MB + Indice: 0.00MB
wp_advads_impressions: Dati: 0.06MB + Indice: 0.00MB
wp_bad_behavior: Dati: 4.02MB + Indice: 0.25MB
wp_commentmeta: Dati: 8.52MB + Indice: 6.95MB
wp_comments: Dati: 47.58MB + Indice: 48.75MB
wp_country: Dati: 0.19MB + Indice: 0.00MB
wp_fv_fp_hls_access_tokens: Dati: 0.02MB + Indice: 0.03MB
wp_links: Dati: 0.02MB + Indice: 0.03MB
wp_no_disposable_email: Dati: 0.02MB + Indice: 0.02MB
wp_options: Dati: 14.19MB + Indice: 2.25MB
wp_postmeta: Dati: 134.61MB + Indice: 54.14MB
wp_posts: Dati: 74.58MB + Indice: 9.47MB
wp_termmeta: Dati: 0.02MB + Indice: 0.03MB
wp_terms: Dati: 0.23MB + Indice: 0.28MB
wp_term_relationships: Dati: 7.52MB + Indice: 6.52MB
wp_term_taxonomy: Dati: 0.23MB + Indice: 0.25MB
wp_usermeta: Dati: 198.84MB + Indice: 170.38MB
wp_users: Dati: 12.48MB + Indice: 13.53MB
wp_wcpdf_credit_note_number: Dati: 0.02MB + Indice: 0.00MB
wp_wcpdf_invoice_number: Dati: 0.02MB + Indice: 0.00MB
wp_wc_download_log: Dati: 0.13MB + Indice: 0.13MB
wp_wc_webhooks: Dati: 0.02MB + Indice: 0.02MB
wp_woocommerce_termmeta: Dati: 0.02MB + Indice: 0.03MB
wp_yith_wcaf_affiliatemeta: Dati: 0.02MB + Indice: 0.03MB
wp_yith_wcaf_affiliates: Dati: 0.16MB + Indice: 0.00MB
wp_yith_wcaf_clicks: Dati: 4.52MB + Indice: 0.00MB
wp_yith_wcaf_commissions: Dati: 0.19MB + Indice: 0.00MB
wp_yith_wcaf_commission_notes: Dati: 0.09MB + Indice: 0.00MB
wp_yith_wcaf_payments: Dati: 0.02MB + Indice: 0.00MB
wp_yith_wcaf_payment_commission: Dati: 0.09MB + Indice: 0.00MB
wp_yith_wcaf_payment_notes: Dati: 0.02MB + Indice: 0.00MB
wp_yith_wcaf_rates: Dati: 0.02MB + Indice: 0.00MB
wp_yoast_seo_links: Dati: 1.47MB + Indice: 0.23MB
wp_yoast_seo_meta: Dati: 0.16MB + Indice: 0.00MB

### Post Type Counts ###

advanced_ads: 29
amn_mi-lite: 2
attachment: 7256
course: 8
course_features: 45
course_promo: 3
csshortcode: 101
dsn_note: 1
lesson: 204
nav_menu_item: 93
newsletter: 8
page: 1198
pcplms_upsell: 7
pcplms_upsell_roles: 5
pcp_survey: 7390
post: 3425
product: 882
revision: 1052
shop_coupon: 81
shop_order: 8050
shop_order_refund: 178
slide: 1
smart_offers: 89
snp_popups: 1
testimonial: 9
wysijap: 1

### Security ###

Secure connection (HTTPS): ❌Il tuo negozio non utilizza HTTPS. Scopri di più su HTTPS e certificati SSL .
Hide errors from visitors: ✔

### Active Plugins (39) ###

User Switching in Admin Bar: by Mark Wilkinson – 1.1.1
Advanced Ads Pro: by Thomas Maier – 2.1.4
Advanced Ads – Responsive Ads: by Thomas Maier – 1.7.3
Advanced Ads – Ad Tracking: by Thomas Maier – 1.8.12
Gestione Ad: by Thomas Maier – 1.8.30
Advanced Cron Manager: by BracketSpace – 2.3.2
Akismet Anti-Spam: by Automattic – 4.0.7
Bad Behavior: by Michael Hampton – 2.2.20
BE Subpages Widget: by Bill Erickson – 1.7
Better Click To Tweet: by Ben Meredith – 5.5
CMS Tree Page View: by Pär Thernström – 1.5
GDPR Cookie Consent: by webtoffee – 1.5.7
Disable admin notification email: by David Silvestri – 0.0.1
Duplicate Post: by Enrico Battocchi – 3.2.2
Enhanced Media Library: by wpUXsolutions – 2.5
Nelio External Featured Image (discontinued): by Nelio Software – 1.4.5
ezPHP: by WebSharks
Inc. (Jason Caldwell) – 160128

FV Player Pro: by Foliovision – 6.6.2
FV Player: by  – 6.6.2
Google Analytics for WordPress by MonsterInsights: by MonsterInsights – 7.0.6
PCP Salvataggio dati questionari: by david silvestri@dot4all S.r.l. – 0.0.2
Prevent Concurrent Logins: by Frankie Jarrett – 0.4.0
Quick Page/Post Redirect Plugin: by anadnet – 5.1.8
Rewrite Rules Inspector: by Daniel Bachhuber
Automattic – 1.2.1

UberMenu Conditionals: by Chris Mavricos
SevenSpark – 3.0

UberMenu 3 - The Ultimate WordPress Mega Menu: by Chris Mavricos
SevenSpark – 3.1

User Role Editor: by Vladimir Garagulya – 4.43
User Switching: by John Blackbourn & contributors – 1.3.1
W3 Total Cache: by Frederick Townes – 0.9.7
WooCommerce PayPal Powered by Braintree Gateway: by WooCommerce – 2.0.4 – Non testato con la versione attiva di WooCommerce
Advanced Order Export For WooCommerce (Pro): by AlgolPlus – 1.5.4
WooCommerce PDF Invoices & Packing Slips: by Ewout Fernhout – 2.1.6
WooCommerce PDF Invoices & Packing Slips Professional: by Ewout Fernhout – 2.2.3
WooCommerce: by Automattic – 3.3.5 – 3.4.2 è disponibile
Yoast SEO: by Team Yoast – 7.6.1
wpMandrill: by Mandrill – 1.33
WP Overnight Sidekick: by Jeremiah Prummer
Ewout Fernhout
Michael Kluver – 2.0.3

PcP Affiliates: by david.silvestri –
YITH WooCommerce Product Bundles Premium: by YITHEMES – 1.2.10

### Settings ###

API Enabled: ✔
Force SSL: ✔
Currency: EUR (€)
Currency Position: right
Thousand Separator: .
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)
yith_bundle (yith_bundle)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)


### WC Pages ###

Shop base: ❌ L'ID della pagina è impostato
ma la pagina non esiste

Carrello: #47192 - /carrello/
Cassa: #15135 - /checkout/
Il mio account: #15148 - /account/
Termini e condizioni: #66032 - /termini/

### Theme ###

Name: PcP
Version: – 0 è disponibile
Author URL: 
Child Theme: ❌ – Se stai modificando WooCommerce o un tema genitore che non hai costruito personalmente
ti consigliamo di utilizzare un child theme. Vedi: Come creare un child theme

WooCommerce Support: ✔

### Templates ###

Archive Template: Il tuo tema ha il file woocommerce.php
non potrai sovrascrivere il template personalizzato woocommerce/archive-product.php poichè woocommerce.php ha priorità su archive-product.php. Questo serve a prevenire problemi di visualizzazione.

Overrides: qf/woocommerce/archive-product.php versione 2.0.0 è obsoleta. La versione attuale è 3.3.0
qf/woocommerce/checkout/form-checkout.php
qf/woocommerce/checkout/form-shipping.php
qf/woocommerce/checkout/payment-method.php
qf/woocommerce/checkout/payment.php versione 2.5.0 è obsoleta. La versione attuale è 3.3.0
qf/woocommerce/checkout/review-order.php versione 2.3.0 è obsoleta. La versione attuale è 3.3.0
qf/woocommerce/checkout/thankyou.php
qf/woocommerce/content-product.php
qf/woocommerce/emails/admin-failed-order.php
qf/woocommerce/emails/customer-completed-order.php
qf/woocommerce/emails/customer-processing-order.php
qf/woocommerce/emails/customer-refunded-order.php
qf/woocommerce/emails/email-downloads.php versione - è obsoleta. La versione attuale è 3.2.0
qf/woocommerce/myaccount/form-edit-account.php versione 2.6.0 è obsoleta. La versione attuale è 3.3.0
qf/woocommerce/myaccount/form-edit-address.php versione 3.0.9 è obsoleta. La versione attuale è 3.3.0
qf/woocommerce/myaccount/form-login.php versione 3.2.0 è obsoleta. La versione attuale è 3.3.0
qf/woocommerce/myaccount/form-lost-password.php versione 3.0.0 è obsoleta. La versione attuale è 3.3.0
qf/woocommerce/myaccount/my-account.php
qf/woocommerce/myaccount/my-orders.php
qf/woocommerce/myaccount/view-order.php

Outdated Templates: ❌Scopri come aggiornare

### Braintree (Credit Card) ###

Environment: Production
Tokenization Enabled: –
Debug Mode: Off

### Braintree (PayPal) ###

Environment: Production
Tokenization Enabled: –
Debug Mode: Off
`
```
</details>

**Comment by mikejolley** [comment]
We're working on improving speed of reports with https://github.com/woocommerce/woocommerce/issues/20243 and some lookup tables. There is no short term solution, but we're targeting a release later this year.

cc @claudiulodro 

Thanks

**Comment by fabiopulitta** [comment]
Hello, Mike.
Any good news about it? I'm having the same issue here. In my case less customers, but more than 100.000 orders. It has been hard days to maintain our e-commerce running.

**Comment by claudiulodro** [comment]
It will be coming as part of [wc-admin](https://woocommerce.wordpress.com/2018/10/18/wc-admin/) in WC 3.6.

**Comment by kurtschlatzer** [comment]
From a recent transaction trace in New Relic:

![screen shot 2019-02-08 at 11 37 06 am](https://user-images.githubusercontent.com/1319939/52491894-248fbf80-2b96-11e9-906a-c20dd04fe947.png)

This is now the average on our site. Is there any way to safely and completely disable reporting in WooCommerce? We don't use it now that we're using our ERP for reports.

**Comment by lkraav** [comment]
@mikejolley this is still a real issue on 3.6.2

Any transaction spike can take the site into completely unresponsive mode, where every `_money_spent` query takes ~1 minute to complete.

Is there really nothing we can do to help MySQL out here?

**Comment by lkraav** [comment]
My incident might actually be caused by @bryceadams Metorik app, where an incoming transaction triggers a webhook to Metorik, which then fires an API request in my direction for its customer profile update, hitting https://github.com/woocommerce/woocommerce/blob/3.6.2/includes/api/v2/class-wc-rest-customers-v2-controller.php#L61 `WC_REST_Customers_V2_Controller::get_formatted_item_data()` for potentially many people in parallel.

Same for `get_order_count()`.

When each of these queries take tens of seconds to complete, we have a problem.

It would be very useful if WC silently (and somehow gracefully, without impacting system performance) indexed everyone in the background.

**Comment by bryceadams** [comment]
@lkraav This is a known issue for us with Woo since the start of Metorik, so the Metorik Helper plugin actually filters the customer responses so it should never try calculate total spent / order count.

**Comment by lkraav** [comment]
@bryceadams thanks for insight.

It actually looks like like this information is getting queried when delivering outbound webhook payload, traceback for `get_order_count` available at https://sentry.io/share/issue/6d8fd864532d4340b9dfd441888ccaf7/

**Comment by bryceadams** [comment]
Ah, interesting! I don't think I filter those so that's good to know - maybe I can add something there. But agree with you that this needs to be improved in WooCommerce core itself.

**Comment by lkraav** [comment]
@bryceadams that patch to Metorik plugin would certainly be most welcome. I'm doing hardcore monkeypatching right now to keep my site alive:

```php
        /**
         * Return the number of orders this customer has.
         *
         * @since 3.0.0
         * @param WC_Customer $customer Customer object.
         * @return integer
         */
        public function get_order_count( &$customer ) {
                $count = get_user_meta( $customer->get_id(), '_order_count', true );

                // @leho 2019.05.09
                // https://github.com/woocommerce/woocommerce/issues/20513
                if ( '' === $count ) {
                        $count = 0;
                }

                if ( '' === $count ) {
                        global $wpdb;

                        $count = $wpdb->get_var(
                                // phpcs:disable WordPress.DB.PreparedSQL.NotPrepared
                                "SELECT COUNT(*)
                                FROM $wpdb->posts as posts
                                LEFT JOIN {$wpdb->postmeta} AS meta ON posts.ID = meta.post_id
        ...
```

**Comment by bryceadams** [comment]
@lkraav maybe try this instead :) https://gist.github.com/bryceadams/050d886159265d6f5e6dbce649552704.

**Comment by lkraav** [comment]
@bryceadams that's solid, but I'm only looking to zero un-indexed (aka new) people.

So in that sense, that filter could check `$value` for whether it's an empty string, and only zero that, right?

**Comment by bryceadams** [comment]
@lkraav ah yep, your idea should do the trick!

**Comment by lkraav** [comment]
https://www.elasticpress.io/blog/2019/05/elasticpress-3-0-released/

> Version 3.0 completely refactors the plugin to use Indexables. In the past, ElasticPress simply integrated with WordPress’s WP_Query API which enabled redirection of post queries through Elasticsearch instead of MySQL. Indexables take this a step further, enabling indexing, search, and querying on anything.

Sounds useful here, maybe, for those that can swing such an integration.

**Comment by fabiopulitta** [comment]
Why is it closed? There's a solution?
Here still the same problem. When the shop reaches 90-100 visitors, it brings me headaches.
Always a lot of "SELECT SUM(meta2.meta_value) FROM wp_posts as posts LEFT JOIN wp_postmeta AS meta ON posts"

**Comment by nhadsall** [comment]
I was so excited to see someone else was having this issue... then I got to the end and saw it was closed? No solution @mikejolley we are not using Metorik, but see this query often brings our site to its knees.

**Comment by mikejolley** [comment]
@nhadsall The WC Admin project is introducing lookup tables for this purpose (coming in WC 4.0). There is no magic solution to make the post-meta version of this query performant unfortunately.

**Comment by bryceadams** [comment]
@nhadsall Best solution is to implement the code I added here - https://gist.github.com/bryceadams/050d886159265d6f5e6dbce649552704 - if you don't need the value calculated and stored inside of Woo. You don't need to use Metorik to use it :)

**Comment by nhadsall** [comment]
Thanks @bryceadams that's a great approach. @mikejolley is there any code partially written for that? I'm thinking it might be possible to create a temporary table that keeps a running total with new orders and and use Bryce's approach to intercept this call?

So, 1. hook into new orders, add total to user's entry in temp table. Then, 2. use a hook to stop the woo code from looking in postmeta.

Although, when I look at the function, it seems to store values in meta already?

https://docs.woocommerce.com/wc-apidocs/source-class-WC_Customer_Data_Store.html#379-422

**Comment by rrennick** [comment]
@nhadsall If you have WooCommerce Admin installed and have imported the analytics data then the rest API has a customer endpoint.

Example from generated test WC data:
```
Request URL: https://example.com/wp-json/wc-analytics/customers/1754
Request Method: GET
Response Time: 0.191 seconds
Response Status: 200 - OK

[
  {
    "id": 1754,
    "user_id": 0,
    "username": "",
    "name": "Cathryn Hintz",
    "email": "",
    "country": "CF",
    "city": "East Tevin",
    "state": "PA",
    "postcode": "37102",
    "date_registered": null,
    "date_last_active": "2018-10-26T02:00:00",
    "date_last_order": "2018-10-25 23:00:00",
    "orders_count": 1,
    "total_spend": 5451.17,
    "avg_order_value": 5451.17,
    "date_registered_gmt": null,
    "date_last_active_gmt": "2018-10-25T23:00:00"
  }
]
```

**Comment by nhadsall** [comment]
So we disabled the query and performance on the DB server started to FLY. It was this single query that brought the site to its knees.

So, I created a table to house order total data and intercepted the query thanks to well placed Woo hooks!

In case anyone else needs it:
1. Create the new table with the query.
2. Add the class
3. Run the query to copy historical data to the new table (FYI this took 25 seconds to copy ALL data, and the woo query that ran would take an average of 20 seconds!)
4. Sit back and watch your site fly.

@mikejolley I'd love to get your feedback on this method. Should I have hooked somewhere else to get updates to order totals?

Thanks to @bryceadams for the excellent idea!

https://gist.github.com/nhadsall/89a71f6c7ca54dcb13063e15977628a2
